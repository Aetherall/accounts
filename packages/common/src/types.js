// @flow

import type { HashAlgorithm } from './config';

export type PasswordType = string | {
  digest: string,
  algorithm: HashAlgorithm
};

export type UserObjectType = {
  username: ?string,
  email: ?string,
  emails?: Object[],
  id: string,
  profile: ?Object,
  services: ?Object
};

export type CreateUserType = {
  username?: string,
  email?: string,
  password?: PasswordType,
  profile?: Object
};

export type LoginUserIdentityType = {
  id?: string,
  username?: string,
  email?: string
};

export type PasswordLoginUserType = string | LoginUserIdentityType;

export type TokensType = {
  accessToken: ?string,
  refreshToken: ?string
};

export type LoginReturnType = {
  sessionId: string,
  user: UserObjectType,
  tokens: TokensType
};

export type ImpersonateReturnType = {
  authorized: boolean,
  tokens?: TokensType,
  user?: UserObjectType
};

export type SessionType = {
  sessionId: string,
  userId: string,
  valid: boolean,
  userAgent: ?string,
  createdAt: string,
  updatedAt: string
};
