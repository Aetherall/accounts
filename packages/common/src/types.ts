import { HashAlgorithm } from './config';

export interface TokenRecord {
  token: string;
  address: string;
  when: number;
  reason: string;
}

export interface EmailRecord {
  address: string;
  verified: boolean;
}

export interface UserObjectType {
  username?: string;
  email?: string;
  emails?: EmailRecord[];
  id: string;
  profile?: object;
  services?: object;
}

export interface CreateUserType {
  username?: string;
  email?: string;
  profile?: object;
}

export interface LoginUserIdentityType {
  id?: string;
  username?: string;
  email?: string;
}

export interface TokensType {
  accessToken?: string;
  refreshToken?: string;
}

export interface LoginReturnType {
  sessionId: string;
  user: UserObjectType;
  tokens: TokensType;
}

export interface ImpersonateReturnType {
  authorized: boolean;
  tokens?: TokensType;
  user?: UserObjectType;
}

export interface SessionType {
  sessionId: string;
  userId: string;
  valid: boolean;
  userAgent?: string;
  createdAt: string;
  updatedAt: string;
}

export type HookListener = (event?: object) => void;
