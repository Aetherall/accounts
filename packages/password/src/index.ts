import AccountsPassword from './accounts-password';

export default AccountsPassword;
export { AccountsPassword };
