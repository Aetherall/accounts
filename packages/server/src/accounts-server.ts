import * as pick from 'lodash/pick';
import * as omit from 'lodash/omit';
import * as isString from 'lodash/isString';
import * as isPlainObject from 'lodash/isPlainObject';
import * as isFunction from 'lodash/isFunction';
import * as find from 'lodash/find';
import * as includes from 'lodash/includes';
import * as get from 'lodash/get';
import { EventEmitter } from 'events';
import * as jwt from 'jsonwebtoken';
import {
  AccountsError,
  toUsernameAndEmail,
  validators,
  UserObjectType,
  CreateUserType,
  LoginReturnType,
  TokensType,
  SessionType,
  ImpersonateReturnType,
  EmailRecord,
  HookListener,
} from '@accounts/common';
import config, { AccountsServerConfiguration } from './config';
import { DBInterface } from './db-interface';
import { verifyPassword, hashPassword, bcryptPassword } from './encryption';
import {
  generateAccessToken,
  generateRefreshToken,
  generateRandomToken,
} from './tokens';
import Email, {
  emailTemplates,
  EmailConnector,
  EmailTemplateType,
} from './email';
import {
  AccountsServerOptions,
  ConnectionInformationsType,
  AuthService,
} from './types';

export interface TokenRecord {
  token: string;
  address: string;
  when: number;
  reason: string;
}

const defaultOptions = {
  tokenSecret: 'secret',
  tokenConfigs: {
    accessToken: {
      expiresIn: '90m',
    },
    refreshToken: {
      expiresIn: '7d',
    },
  },
  emailTemplates,
  userObjectSanitizer: (user: UserObjectType) => user,
};

export type RemoveListnerHandle = () => EventEmitter;

export const ServerHooks = {
  LoginSuccess: 'LoginSuccess',
  LoginError: 'LoginError',
  LogoutSuccess: 'LogoutSuccess',
  LogoutError: 'LogoutError',
  CreateUserSuccess: 'CreateUserSuccess',
  CreateUserError: 'CreateUserError',
  ResumeSessionSuccess: 'ResumeSessionSuccess',
  ResumeSessionError: 'ResumeSessionError',
  RefreshTokensSuccess: 'RefreshTokensSuccess',
  RefreshTokensError: 'RefreshTokensError',
  ImpersonationSuccess: 'ImpersonationSuccess',
  ImpersonationError: 'ImpersonationError',
};

export class AccountsServer {
  // TODO private services: AuthService[];
  public options: AccountsServerOptions;
  public email: EmailConnector;
  private services: any;
  private db: DBInterface;
  private hooks: EventEmitter;

  constructor(options: AccountsServerOptions, services: any) {
    this.options = { ...defaultOptions, ...options };
    if (!this.options.db) {
      throw new AccountsError('A database driver is required');
    }
    // TODO if this.options.tokenSecret === 'secret' warm user to change it

    this.services = services;
    this.db = this.options.db;

    // Set the db to all services
    // tslint:disable-next-line
    for (const service in this.services) {
      this.services[service].db = this.db;
      this.services[service].server = this;
    }

    // Initialize emails
    // this.email = this.options.sendMail
    //   ? { sendMail: this.options.sendMail }
    //   : new Email(this.options.email);

    // Initialize hooks
    this.hooks = new EventEmitter();
  }

  public onLoginSuccess(callback: HookListener): RemoveListnerHandle {
    return this.on(ServerHooks.LoginSuccess, callback);
  }

  public onLoginError(callback: HookListener): RemoveListnerHandle {
    return this.on(ServerHooks.LoginError, callback);
  }

  public onLogoutSuccess(callback: HookListener): RemoveListnerHandle {
    return this.on(ServerHooks.LogoutSuccess, callback);
  }

  public onLogoutError(callback: HookListener): RemoveListnerHandle {
    return this.on(ServerHooks.LogoutError, callback);
  }

  public onCreateUserSuccess(callback: HookListener): RemoveListnerHandle {
    return this.on(ServerHooks.CreateUserSuccess, callback);
  }

  public onCreateUserError(callback: HookListener): RemoveListnerHandle {
    return this.on(ServerHooks.CreateUserError, callback);
  }

  public onResumeSessionSuccess(callback: HookListener): RemoveListnerHandle {
    return this.on(ServerHooks.ResumeSessionSuccess, callback);
  }

  public onResumeSessionError(callback: HookListener): RemoveListnerHandle {
    return this.on(ServerHooks.ResumeSessionError, callback);
  }

  public onRefreshTokensSuccess(callback: HookListener): RemoveListnerHandle {
    return this.on(ServerHooks.RefreshTokensSuccess, callback);
  }

  public onRefreshTokensError(callback: HookListener): RemoveListnerHandle {
    return this.on(ServerHooks.RefreshTokensError, callback);
  }

  public onImpersonationSuccess(callback: HookListener): RemoveListnerHandle {
    return this.on(ServerHooks.ImpersonationSuccess, callback);
  }

  public onImpersonationError(callback: HookListener): RemoveListnerHandle {
    return this.on(ServerHooks.ImpersonationError, callback);
  }

  public async loginWithService(
    serviceName: string,
    params,
    infos: ConnectionInformationsType
  ): Promise<LoginReturnType> {
    if (!this.services[serviceName]) {
      throw new Error(
        `No service with the name ${serviceName} was registered.`
      );
    }
    const user: UserObjectType = await this.services[serviceName].authenticate(
      params
    );
    if (!user) {
      throw new Error(
        `Service ${serviceName} was not able to authenticate user`
      );
    }
    return this.loginWithUser(user, infos);
  }

  /**
   * @description Server use only. This method creates a session
   *              without authenticating any user identity.
   *              Any authentication should happen before calling this function.
   * @param {UserObjectType} userId - The user object.
   * @param {string} ip - User's ip.
   * @param {string} userAgent - User's client agent.
   * @returns {Promise<LoginReturnType>} - Session tokens and user object.
   */
  public async loginWithUser(
    user: UserObjectType,
    infos: ConnectionInformationsType
  ): Promise<LoginReturnType> {
    const { ip, userAgent } = infos;
    const sessionId = await this.db.createSession(user.id, ip, userAgent);
    const { accessToken, refreshToken } = this.createTokens(sessionId);

    const loginResult = {
      sessionId,
      user: this.sanitizeUser(user),
      tokens: {
        refreshToken,
        accessToken,
      },
    };

    return loginResult;
  }

  /**
   * @description Impersonate to another user.
   * @param {string} accessToken - User access token.
   * @param {string} username - impersonated user username.
   * @param {string} ip - The user ip.
   * @param {string} userAgent - User user agent.
   * @returns {Promise<Object>} - ImpersonateReturnType
   */
  public async impersonate(
    accessToken: string,
    username: string,
    ip: string,
    userAgent: string
  ): Promise<ImpersonateReturnType> {
    try {
      if (!isString(accessToken)) {
        throw new AccountsError('An access token is required');
      }

      try {
        jwt.verify(accessToken, this.options.tokenSecret);
      } catch (err) {
        throw new AccountsError('Access token is not valid');
      }

      const session = await this.findSessionByAccessToken(accessToken);

      if (!session.valid) {
        throw new AccountsError('Session is not valid for user');
      }

      const user = await this.db.findUserById(session.userId);

      if (!user) {
        throw new AccountsError('User not found');
      }

      const impersonatedUser = await this.db.findUserByUsername(username);
      if (!impersonatedUser) {
        throw new AccountsError(`User ${username} not found`);
      }

      if (!this.options.impersonationAuthorize) {
        return { authorized: false };
      }

      const isAuthorized = await this.options.impersonationAuthorize(
        user,
        impersonatedUser
      );

      if (!isAuthorized) {
        return { authorized: false };
      }

      const newSessionId = await this.db.createSession(
        impersonatedUser.id,
        ip,
        userAgent,
        { impersonatorUserId: user.id }
      );
      const impersonationTokens = this.createTokens(newSessionId, true);
      const impersonationResult = {
        authorized: true,
        tokens: impersonationTokens,
        user: this.sanitizeUser(impersonatedUser),
      };

      this.hooks.emit(
        ServerHooks.ImpersonationSuccess,
        user,
        impersonationResult
      );

      return impersonationResult;
    } catch (e) {
      this.hooks.emit(ServerHooks.ImpersonationError, e);

      throw e;
    }
  }

  /**
   * @description Refresh a user token.
   * @param {string} accessToken - User access token.
   * @param {string} refreshToken - User refresh token.
   * @param {string} ip - User ip.
   * @param {string} userAgent - User user agent.
   * @returns {Promise<Object>} - LoginReturnType.
   */
  public async refreshTokens(
    accessToken: string,
    refreshToken: string,
    ip: string,
    userAgent: string
  ): Promise<LoginReturnType> {
    try {
      if (!isString(accessToken) || !isString(refreshToken)) {
        throw new AccountsError('An accessToken and refreshToken are required');
      }

      let sessionId;
      try {
        jwt.verify(refreshToken, this.options.tokenSecret);
        const decodedAccessToken: any = jwt.verify(
          accessToken,
          this.options.tokenSecret,
          {
            ignoreExpiration: true,
          }
        );
        sessionId = decodedAccessToken.data.sessionId;
      } catch (err) {
        throw new AccountsError('Tokens are not valid');
      }

      const session: SessionType = await this.db.findSessionById(sessionId);
      if (!session) {
        throw new AccountsError('Session not found');
      }

      if (session.valid) {
        const user = await this.db.findUserById(session.userId);
        if (!user) {
          throw new AccountsError('User not found', { id: session.userId });
        }
        const tokens = this.createTokens(sessionId);
        await this.db.updateSession(sessionId, ip, userAgent);

        const result = {
          sessionId,
          user: this.sanitizeUser(user),
          tokens,
        };

        this.hooks.emit(ServerHooks.RefreshTokensSuccess, result);

        return result;
      } else {
        throw new AccountsError('Session is no longer valid', {
          id: session.userId,
        });
      }
    } catch (err) {
      this.hooks.emit(ServerHooks.RefreshTokensError, err);

      throw err;
    }
  }

  /**
   * @description Refresh a user token.
   * @param {string} sessionId - User session id.
   * @param {boolean} isImpersonated - Should be true if impersonating another user.
   * @returns {Promise<Object>} - Return a new accessToken and refreshToken.
   */
  public createTokens(
    sessionId: string,
    isImpersonated: boolean = false
  ): TokensType {
    const {
      tokenSecret = config.tokenSecret,
      tokenConfigs = config.tokenConfigs,
    } = this.options;
    const accessToken = generateAccessToken({
      data: {
        sessionId,
        isImpersonated,
      },
      secret: tokenSecret,
      config: tokenConfigs.accessToken || {},
    });
    const refreshToken = generateRefreshToken({
      secret: tokenSecret,
      config: tokenConfigs.refreshToken || {},
    });
    return { accessToken, refreshToken };
  }

  /**
   * @description Logout a user and invalidate his session.
   * @param {string} accessToken - User access token.
   * @returns {Promise<void>} - Return a promise.
   */
  public async logout(accessToken: string): Promise<void> {
    try {
      const session: SessionType = await this.findSessionByAccessToken(
        accessToken
      );

      if (session.valid) {
        const user = await this.db.findUserById(session.userId);

        if (!user) {
          throw new AccountsError('User not found', { id: session.userId });
        }

        await this.db.invalidateSession(session.sessionId);
        this.hooks.emit(
          ServerHooks.LogoutSuccess,
          this.sanitizeUser(user),
          session,
          accessToken
        );
      } else {
        throw new AccountsError('Session is no longer valid', {
          id: session.userId,
        });
      }
    } catch (error) {
      this.hooks.emit(ServerHooks.LogoutError, error);

      throw error;
    }
  }

  public async resumeSession(accessToken: string): Promise<UserObjectType> {
    try {
      const session: SessionType = await this.findSessionByAccessToken(
        accessToken
      );

      if (session.valid) {
        const user = await this.db.findUserById(session.userId);

        if (!user) {
          throw new AccountsError('User not found', { id: session.userId });
        }

        if (this.options.resumeSessionValidator) {
          try {
            await this.options.resumeSessionValidator(user, session);
          } catch (e) {
            throw new AccountsError(e, { id: session.userId }, 403);
          }
        }

        this.hooks.emit(ServerHooks.ResumeSessionSuccess, user, accessToken);

        return this.sanitizeUser(user);
      }

      this.hooks.emit(
        ServerHooks.ResumeSessionError,
        new AccountsError('Invalid Session', { id: session.userId })
      );

      return null;
    } catch (e) {
      this.hooks.emit(ServerHooks.ResumeSessionError, e);

      throw e;
    }
  }

  public async findSessionByAccessToken(
    accessToken: string
  ): Promise<SessionType> {
    if (!isString(accessToken)) {
      throw new AccountsError('An accessToken is required');
    }

    let sessionId;
    try {
      const decodedAccessToken: any = jwt.verify(
        accessToken,
        this.options.tokenSecret
      );
      sessionId = decodedAccessToken.data.sessionId;
    } catch (err) {
      throw new AccountsError('Tokens are not valid');
    }

    const session: SessionType = await this.db.findSessionById(sessionId);
    if (!session) {
      throw new AccountsError('Session not found');
    }

    return session;
  }

  /**
   * @description Find a user by his id.
   * @param {string} userId - User id.
   * @returns {Promise<Object>} - Return a user or null if not found.
   */
  public findUserById(userId: string): Promise<UserObjectType> {
    return this.db.findUserById(userId);
  }

  /**
   * @description Change the profile for a user.
   * @param {string} userId - User id.
   * @param {Object} profile - The new user profile.
   * @returns {Promise<void>} - Return a Promise.
   */
  public async setProfile(userId: string, profile: object): Promise<void> {
    const user = await this.db.findUserById(userId);
    if (!user) {
      throw new AccountsError('User not found', { id: userId });
    }
    await this.db.setProfile(userId, profile);
  }

  /**
   * @description Update the profile for a user,
   * the new profile will be added to the existing one.
   * @param {string} userId - User id.
   * @param {Object} profile - User profile to add.
   * @returns {Promise<Object>} - Return a Promise.
   */
  public async updateProfile(userId: string, profile: object): Promise<object> {
    const user = await this.db.findUserById(userId);
    if (!user) {
      throw new AccountsError('User not found', { id: userId });
    }
    return this.db.setProfile(userId, { ...user.profile, ...profile });
  }

  public on(eventName: string, callback: HookListener): RemoveListnerHandle {
    this.hooks.on(eventName, callback);

    return () => this.hooks.removeListener(eventName, callback);
  }

  public isTokenExpired(token: string, tokenRecord?: TokenRecord): boolean {
    return (
      !tokenRecord ||
      Number(tokenRecord.when) + this.options.emailTokensExpiry < Date.now()
    );
  }

  public prepareMail(
    to: string,
    token: string,
    user: UserObjectType,
    pathFragment: string,
    emailTemplate: EmailTemplateType,
    from: string
  ): any {
    if (this.options.prepareMail) {
      return this.options.prepareMail(
        to,
        token,
        user,
        pathFragment,
        emailTemplate,
        from
      );
    }
    return this.defaultPrepareEmail(
      to,
      token,
      user,
      pathFragment,
      emailTemplate,
      from
    );
  }

  public sanitizeUser(user: UserObjectType): UserObjectType {
    const { userObjectSanitizer } = this.options;

    return userObjectSanitizer(this.internalUserSanitizer(user), omit, pick);
  }

  private internalUserSanitizer(user: UserObjectType): UserObjectType {
    return omit(user, ['services']);
  }

  private defaultPrepareEmail(
    to: string,
    token: string,
    user: UserObjectType,
    pathFragment: string,
    emailTemplate: EmailTemplateType,
    from: string
  ): object {
    const tokenizedUrl = this.defaultCreateTokenizedUrl(pathFragment, token);
    return {
      from: emailTemplate.from || from,
      to,
      subject: emailTemplate.subject(user),
      text: emailTemplate.text(user, tokenizedUrl),
    };
  }

  private defaultCreateTokenizedUrl(
    pathFragment: string,
    token: string
  ): string {
    const siteUrl = this.options.siteUrl || config.siteUrl;
    return `${siteUrl}/${pathFragment}/${token}`;
  }
}

export default AccountsServer;
