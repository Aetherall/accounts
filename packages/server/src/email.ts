import email, { Client } from 'emailjs';
import { UserObjectType } from '@accounts/common';

export interface EmailTemplateType {
  from?: string;
  subject: (user?: UserObjectType) => string;
  text: (user: UserObjectType, url: string) => string;
}

export interface EmailTemplatesType {
  from: string;
  verifyEmail: EmailTemplateType;
  resetPassword: EmailTemplateType;
  enrollAccount: EmailTemplateType;
}

export const emailTemplates = {
  from: 'js-accounts <no-reply@js-accounts.com>',

  verifyEmail: {
    subject: () => 'Verify your account email',
    text: (user: UserObjectType, url: string) =>
      `To verify your account email please click on this link: ${url}`,
  },

  resetPassword: {
    subject: () => 'Reset your password',
    text: (user: UserObjectType, url: string) =>
      `To reset your password please click on this link: ${url}`,
  },

  enrollAccount: {
    subject: () => 'Set your password',
    text: (user: UserObjectType, url: string) =>
      `To set your password please click on this link: ${url}`,
  },
};

export interface EmailConnector {
  sendMail(mail: object): Promise<object>;
}

class Email {
  private server: Client;

  constructor(emailConfig: object) {
    if (emailConfig) {
      this.server = email.server.connect(emailConfig);
    }
  }

  public sendMail(mail: object): Promise<object> {
    return new Promise((resolve, reject) => {
      // If no configuration for email just warn the user
      if (!this.server) {
        // tslint:disable-next-line no-console
        console.warn(
          'No configuration for email, you must set an email configuration'
        );
        resolve();
        return;
      }
      this.server.send(mail, (err: object, message: object) => {
        if (err) {
          return reject(err);
        }
        return resolve(message);
      });
    });
  }
}

export default Email;
