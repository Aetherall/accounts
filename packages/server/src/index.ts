import { AccountsServer } from './accounts-server';
import * as encryption from './encryption';
import config from './config';
import { generateRandomToken } from './tokens';
import { DBInterface } from './types';

export { AccountsServer, encryption, config, DBInterface, generateRandomToken };
