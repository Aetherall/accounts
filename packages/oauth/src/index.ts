import { AccountsServer, DBInterface } from '@accounts/server';
import * as requestPromise from 'request-promise';

export class AccountsOauth {
  private options: any;
  private db: DBInterface;
  private accountsServer: AccountsServer;

  constructor(options) {
    this.options = options;
  }

  public async authenticate(params: any): Promise<any> {
    if (!params.provider || !this.options[params.provider]) {
      throw new Error('Invalid provider');
    }

    const { userProvider = this[params.provider] } = this.options[
      params.provider
    ];

    if (!userProvider) {
      throw new Error('Invalid provider');
    }

    const oauthUser = await userProvider(params);
    let user = await this.db.findUserByServiceId(params.provider, oauthUser.id);

    if (!user && oauthUser.email) {
      user = await this.db.findUserByEmail(oauthUser.email);
    }

    if (!user) {
      const userId = await this.db.createUser({
        email: oauthUser.email,
        profile: oauthUser.profile,
      });

      user = await this.db.findUserById(userId);
    } else {
      // If user exist, attmpt to update profile
      this.db.setProfile(user.id, oauthUser.profile);
    }
    await this.db.setService(user.id, params.provider, oauthUser);
    return user;
  }

  private async facebook(params): Promise<any> {
    if (!params.access_token) {
      throw new Error('No access token provided');
    }
    const user = await requestPromise({
      method: 'GET',
      json: true,
      uri: 'https://graph.facebook.com/v2.9/me',
      qs: { access_token: params.access_token },
    });
    // TODO get user email and save it
    return {
      id: user.id,
      name: user.name,
    };
  }
}
