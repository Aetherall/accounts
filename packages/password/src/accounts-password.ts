import { trim, isEmpty, isFunction, isString, isPlainObject } from 'lodash';
import {
  CreateUserType,
  UserObjectType,
  HashAlgorithm,
  LoginUserIdentityType,
} from '@accounts/common';
import { DBInterface } from '@accounts/server';
import { hashPassword, bcryptPassword, verifyPassword } from './encryption';
import {
  PasswordCreateUserType,
  PasswordLoginType,
  PasswordType,
} from './types';

export const isEmail = (email?: string) => {
  const re = /^(([^<>()[\]\\.,;:\s@\"]+(\.[^<>()[\]\\.,;:\s@\"]+)*)|(\".+\"))@((\[[0-9]{1,3}\.[0-9]{1,3}\.[0-9]{1,3}\.[0-9]{1,3}\])|(([a-zA-Z\-0-9]+\.)+[a-zA-Z]{2,}))$/;
  return email && re.test(email);
};

export interface AccountsPasswordOptions {
  passwordHashAlgorithm?: HashAlgorithm;
  passwordResetTokenExpirationInDays?: number;
  passwordEnrollTokenExpirationInDays?: number;
  minimumPasswordLength?: number;
  validateNewUser?: (user: CreateUserType) => Promise<boolean>;
  validateEmail?(email?: string): boolean;
  validatePassword?(email?: string): boolean;
  validateUsername?(email?: string): boolean;
}

const defaultOptions = {
  passwordResetTokenExpirationInDays: 3,
  passwordEnrollTokenExpirationInDays: 30,
  minimumPasswordLength: 7,
  validateEmail(email?: string): boolean {
    const isValid = !isEmpty(trim(email || '')) && isEmail(email);
    return Boolean(isValid);
  },
  validatePassword(password?: string): boolean {
    const isValid = !isEmpty(password);
    return isValid;
  },
  validateUsername(username?: string): boolean {
    const usernameRegex = /^[a-zA-Z][a-zA-Z0-9]*$/;
    const isValid =
      username && !isEmpty(trim(username)) && usernameRegex.test(username);
    return Boolean(isValid);
  },
};

export default class AccountsPassword {
  private serviceName: string;
  private options: AccountsPasswordOptions;
  private db: DBInterface;

  constructor(options: AccountsPasswordOptions) {
    this.options = { ...defaultOptions, ...options };
  }

  public async authenticate(
    params: PasswordLoginType
  ): Promise<UserObjectType> {
    const { user, password } = params;
    if (!user || !password) {
      throw new Error('Unrecognized options for login request');
    }
    if ((!isString(user) && !isPlainObject(user)) || !isString(password)) {
      throw new Error('Match failed');
    }

    let foundUser;
    /* if (this._options.passwordAuthenticator) {
      foundUser = await this._externalPasswordAuthenticator(
        this._options.passwordAuthenticator,
        user,
        password
      );
    } else { */
    foundUser = await this.passwordAuthenticator(user, password);

    if (!foundUser) {
      throw new Error('User not found');
    }

    return foundUser;
  }

  /**
   * @description Create a new user.
   * @param user - The user object.
   * @returns Return the id of user created.
   */
  public async createUser(user: PasswordCreateUserType): Promise<string> {
    if (
      !this.options.validateUsername(user.username) &&
      !this.options.validateEmail(user.email)
    ) {
      throw new Error('Username or Email is required');
    }

    if (user.username && (await this.db.findUserByUsername(user.username))) {
      throw new Error('Username already exists');
    }

    if (user.email && (await this.db.findUserByEmail(user.email))) {
      throw new Error('Email already exists');
    }

    let password;
    if (user.password) {
      password = await this.hashAndBcryptPassword(user.password);
    }

    const proposedUserObject = {
      username: user.username,
      email: user.email && user.email.toLowerCase(),
      password,
      profile: user.profile,
    };

    const { validateNewUser } = this.options;
    if (
      isFunction(validateNewUser) &&
      !await validateNewUser(proposedUserObject)
    ) {
      throw new Error('User invalid');
    }

    return this.db.createUser(proposedUserObject);
  }

  private async passwordAuthenticator(
    user: string | LoginUserIdentityType,
    password: PasswordType
  ): Promise<any> {
    const { username, email, id } = isString(user)
      ? this.toUsernameAndEmail({ user })
      : this.toUsernameAndEmail({ ...user });

    let foundUser;

    if (id) {
      // this._validateLoginWithField('id', user);
      foundUser = await this.db.findUserById(id);
    } else if (username) {
      // this._validateLoginWithField('username', user);
      foundUser = await this.db.findUserByUsername(username);
    } else if (email) {
      // this._validateLoginWithField('email', user);
      foundUser = await this.db.findUserByEmail(email);
    }

    if (!foundUser) {
      throw new Error('User not found');
    }

    const hash = foundUser.service[this.serviceName].bcrypt;
    if (!hash) {
      throw new Error('User has no password set');
    }

    const hashAlgorithm = this.options.passwordHashAlgorithm;
    const pass: any = hashAlgorithm
      ? hashPassword(password, hashAlgorithm)
      : password;
    const isPasswordValid = await verifyPassword(pass, hash);

    if (!isPasswordValid) {
      throw new Error('Incorrect password');
    }

    return foundUser;
  }

  private async hashAndBcryptPassword(password: PasswordType): Promise<string> {
    const hashAlgorithm = this.options.passwordHashAlgorithm;
    const hashedPassword: any = hashAlgorithm
      ? hashPassword(password, hashAlgorithm)
      : password;
    return bcryptPassword(hashedPassword);
  }

  /**
   * Given a username, user and/or email figure out the username and/or email.
   *
   * @param user An object containing at least `username`, `user` and/or `email`.
   * @returns An object containing `id`, `username` and `email`.
   */
  private toUsernameAndEmail({ user, username, email, id }: any): any {
    if (user && !username && !email) {
      if (isEmail(user)) {
        email = user;
        username = null;
      } else {
        username = user;
        email = null;
      }
    }
    return { username, email, id };
  }
}
