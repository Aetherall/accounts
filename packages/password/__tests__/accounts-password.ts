import { AccountsPassword } from '../src';

describe('AccountsPassword', () => {
  describe('config', () => {
    it('should have default options', async () => {
      const password = new AccountsPassword({});
      expect(password.options.passwordResetTokenExpirationInDays).toBe(3);
    });
  });
});
