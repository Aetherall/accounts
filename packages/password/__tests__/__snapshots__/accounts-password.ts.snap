// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`AccountsPassword addEmail call this.db.addEmail 1`] = `
Array [
  "id",
  "email",
  true,
]
`;

exports[`AccountsPassword authenticate return user 1`] = `"User not found"`;

exports[`AccountsPassword authenticate throws on invalid params 1`] = `"Unrecognized options for login request"`;

exports[`AccountsPassword authenticate throws on invalid type params 1`] = `"Match failed"`;

exports[`AccountsPassword findUserByEmail call this.db.findUserByEmail 1`] = `
Array [
  "email",
]
`;

exports[`AccountsPassword findUserByUsername call this.db.findUserByUsername 1`] = `
Array [
  "email",
]
`;

exports[`AccountsPassword removeEmail call this.db.removeEmail 1`] = `
Array [
  "id",
  "email",
]
`;

exports[`AccountsPassword resetPassword reset password and invalidate all sessions 1`] = `
Array [
  "id",
]
`;

exports[`AccountsPassword resetPassword throws when token have invalid email 1`] = `"Token has invalid email address"`;

exports[`AccountsPassword resetPassword throws when token is expired 1`] = `"Reset password link expired"`;

exports[`AccountsPassword resetPassword throws when token not found 1`] = `"Reset password link expired"`;

exports[`AccountsPassword sendEnrollmentEmail throws if user is not found 1`] = `"User not found"`;

exports[`AccountsPassword sendResetPasswordEmail throws if user is not found 1`] = `"User not found"`;

exports[`AccountsPassword sendVerificationEmail throws if user is not found 1`] = `"User not found"`;

exports[`AccountsPassword sendVerificationEmail throws when invalid address 1`] = `"No such email address for user"`;

exports[`AccountsPassword verifyEmail call this.db.verifyEmail 1`] = `
Array [
  undefined,
  "john.doe@gmail.com",
]
`;

exports[`AccountsPassword verifyEmail throws when email not found 1`] = `"Verify email link is for unknown address"`;

exports[`AccountsPassword verifyEmail throws when no user is found 1`] = `"Verify email link expired"`;

exports[`AccountsPassword verifyEmail throws when token not found 1`] = `"Verify email link expired"`;
